opam2: fails to recompile package due to "not removing files that changed since"

I don't have small repro at hand currently, here are the excerpts from the build log, maybe that rings a bell immediately : 
```
[#] opam upgrade
The following actions will be performed:
  - upgrade   pcre       7.3.0 to 7.3.3
  - recompile ocamlnet   4.1.4          [uses pcre]
...

[WARNING] While removing ocamlnet.4.1.4: not removing files that changed since:
            - lib/netstring-pcre/netstring_pcre.mli
            - lib/netstring-pcre/netstring_pcre.cmi
            - lib/netstring-pcre/netstring-pcre.p.cmxa
            - lib/netstring-pcre/netstring-pcre.p.a
            - lib/netstring-pcre/netstring-pcre.cmxa
            - lib/netstring-pcre/netstring-pcre.cma
            - lib/netstring-pcre/netstring-pcre.a
            - lib/netstring-pcre/META
 
[NOTE] While removing ocamlnet.4.1.4: not removing non-empty directories:
         - lib/netstring-pcre
 
-> removed   ocamlnet.4.1.4
-> removed   pcre.7.3.0
-> installed pcre.7.3.3
[ERROR] The installation of ocamlnet failed at "make install".
Processing 12/18: [ocamlnet: ocamlfind remove]
#=== ERROR while installing ocamlnet.4.1.4 ====================================#
# context      2.0.0~beta4 | linux/x86_64 | ocaml-base-compiler.4.05.0 | https://opam.ocaml.org/2.0#2017-10-20 20:08
# path         ~/.opam/4.05.0/.opam-switch/build/ocamlnet.4.1.4
# command      make install
# exit-code    2
# env-file     ~/.opam/log/ocamlnet-7470-57f057.env
# output-file  ~/.opam/log/ocamlnet-7470-57f057.out
### output ###
# [...]
# files=`../../tools/collect_files *.mli *.cmi *.cma *.cmxa *.a dll* META ` && \
# ocamlfind install netstring-pcre $files
# ocamlfind: Package netstring-pcre is already installed
#  - (file /home/user/.opam/4.05.0/lib/netstring-pcre/META already exists)
# ../../Makefile.rules:163: recipe for target 'install-findlib' failed
# make[2]: *** [install-findlib] Error 2
# make[2]: Leaving directory '/home/user/.opam/4.05.0/.opam-switch/build/ocamlnet.4.1.4/src/netstring-pcre'
# ../../Makefile.rules:148: recipe for target 'install' failed
# make[1]: *** [install] Error 2
# make[1]: Leaving directory '/home/user/.opam/4.05.0/.opam-switch/build/ocamlnet.4.1.4/src/netstring-pcre'
# Makefile:58: recipe for target 'install' failed
# make: *** [install] Error 2
```

--- a/src/client/opamSolution.ml
+++ b/src/client/opamSolution.ml
@@ -414,8 +414,14 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
       let t0 = Unix.gettimeofday () in
       fun () -> Hashtbl.add timings action (Unix.gettimeofday () -. t0)
     in
+    let not_yet_removed =
+      PackageActionGraph.fold_descendants (function
+          | `Remove nv -> OpamPackage.Set.add nv
+          | _ -> fun acc -> acc)
+        OpamPackage.Set.empty action_graph action
+    in
     let visible_installed =
-      OpamPackage.Set.Op.(minimal_install ++ installed)
+      OpamPackage.Set.Op.(minimal_install ++ not_yet_removed ++ installed)
     in
     let t =
       { !t_ref with
@@ -433,7 +439,7 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
         OpamConsole.msg "Faking installation of %s\n"
           (OpamPackage.to_string nv);
         add_to_install nv;
-        Done (`Successful (OpamPackage.Set.add nv installed , removed))
+        Done (`Successful (OpamPackage.Set.add nv installed, removed))
       | `Remove nv ->
         remove_from_install nv;
         Done (`Successful (installed, OpamPackage.Set.add nv removed))

--- a/src/client/opamSolution.ml
+++ b/src/client/opamSolution.ml
@@ -415,10 +415,13 @@ let parallel_apply t _action ~requested ?add_roots action_graph =
       fun () -> Hashtbl.add timings action (Unix.gettimeofday () -. t0)
     in
     let not_yet_removed =
-      PackageActionGraph.fold_descendants (function
-          | `Remove nv -> OpamPackage.Set.add nv
-          | _ -> fun acc -> acc)
-        OpamPackage.Set.empty action_graph action
+      match action with
+      | `Remove _ ->
+        PackageActionGraph.fold_descendants (function
+            | `Remove nv -> OpamPackage.Set.add nv
+            | _ -> fun acc -> acc)
+          OpamPackage.Set.empty action_graph action
+      | _ -> OpamPackage.Set.empty
     in
     let visible_installed =
       OpamPackage.Set.Op.(minimal_install ++ not_yet_removed ++ installed)

--- a/src/solver/opamActionGraph.ml
+++ b/src/solver/opamActionGraph.ml
@@ -134,6 +134,7 @@ module type SIG = sig
   include OpamParallel.GRAPH with type V.t = package OpamTypes.action
   val reduce: t -> t
   val explicit: ?noop_remove:(package -> bool) -> t -> t
+  val fold_descendants: (V.t -> 'a -> 'a) -> 'a -> t -> V.t -> 'a
 end
 
 module Make (A: ACTION) : SIG with type package = A.package = struct
@@ -266,4 +267,7 @@ module Make (A: ACTION) : SIG with type package = A.package = struct
         | `Install _ | `Reinstall _ | `Change _ | `Build _ -> ())
       g;
     g
+
+  let rec fold_descendants f acc t v =
+    fold_succ (fun v acc -> fold_descendants f acc t v) t v (f v acc)
 end

--- a/src/solver/opamActionGraph.mli
+++ b/src/solver/opamActionGraph.mli
@@ -43,7 +43,12 @@ module type SIG = sig
       The argument [noop_remove] is a function that should return `true`
       for package where the `remove` action is known not to modify the
       filesystem (such as `conf-*` package). *)
-  val explicit: ?noop_remove:(package -> bool) -> t -> t end
+  val explicit: ?noop_remove:(package -> bool) -> t -> t
+
+  (** Folds on all recursive successors of the given action, including itself,
+      depth-first. Assumes an acyclic graph. *)
+  val fold_descendants: (V.t -> 'a -> 'a) -> 'a -> t -> V.t -> 'a
+end
 
 module Make (A: ACTION) : SIG with type package = A.package
 

